build/link-wrapper.sh

```diff
@@ -20,13 +20,23 @@ case $1 in
 
 	OPTIONS=$( eval echo "$OPTIONS $LDFLAGS \$$2_OBJECTS \$$2_LIBS \$$2_STATIC_LIBS_REAL \$$2_LDFLAGS -o $3" )
 
-	if echo $OPTIONS | grep -q -- -static-libstdc\+\+ ; then
-	    OPTIONS=$( echo $OPTIONS | sed -e s,-static-libstdc\+\+,, -e s,\$,\ `$CC -print-file-name=libstdc++.a`\ -lm, )
+	if [[ "${OPTIONS}" == *"-static-libstdc++"* ]]; then
+	  staticlibs=()
+		staticpaths=''
+	  if [[ "${CC}" == *"gcc"* ]]; then
+	    staticlibs+=('libstdc++.a')
+	  elif [[ "${CXX}" == *"clang++"* ]]; then
+	    staticlibs+=('libc++.a' 'libc++abi.a')
```

Comment:
These are two different configurations.  You can use clang with libstdc++, and you can use clang with libc++.  If anyone was compiling with clang on linux via `make CC=clang CXX=clang++`, then you're likely going to break them.

Also note that #1862 means that compiling like this might not result in a working executable.

Reply:
These are meant to specify 2 separate configurations. I did use 2 different compiler variables in order to be able to identify the different configurations more easy. The first tests for gcc and the second test for compiling c++ source using clang. There are link problems when linking clang against our current code using libstdc++ that I have not been able to resolve. I could only get clang to work when linking against libc++.

Reply:
To be clear, the link problem of clang using libstdc++ is a duplicated symbol defined in two libraries. It seems linking with libstdc++ is not working for clang.

Reply:
So I changed clang to link 'libstdc++.a', and got errors during linking (different from what @AlvinMooreSr showed me the other day, which is the duplicated symbol error):

```
Linking        bin/fdbserver
ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by vector:1010 (/usr/local/include/c++/v1/vector:1010)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(applyMetadataMutations(UID const&, Arena&, VectorRef<MutationRef, (VecSerStrategy)0> const&, IKeyValueStore*, LogPushData*, bool*, Reference<ILogSystem>, long, KeyRangeMap<std::__1::set<Standalone<StringRef>, std::__1::less<Standalone<StringRef> >, std::__1::allocator<Standalone<StringRef> > >, int, ConstantMetric<int> >*, KeyRangeMap<ServerCacheInfo, int, ConstantMetric<int> >*, std::__1::map<Standalone<StringRef>, applyMutationsData, std::__1::less<Standalone<StringRef> >, std::__1::allocator<std::__1::pair<Standalone<StringRef> const, applyMutationsData> > >*, RequestStream<CommitTransactionRequest>, Database, NotifiedVersion*, std::__1::map<UID, Reference<StorageInfo>, std::__1::less<UID>, std::__1::allocator<std::__1::pair<UID const, Reference<StorageInfo> > > >*, std::__1::map<Tag, long, std::__1::less<Tag>, std::__1::allocator<std::__1::pair<Tag const, long> > >*, bool))

ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by vector:1010 (/usr/local/include/c++/v1/vector:1010)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(applyMetadataMutations(UID const&, Arena&, VectorRef<MutationRef, (VecSerStrategy)0> const&, IKeyValueStore*, LogPushData*, bool*, Reference<ILogSystem>, long, KeyRangeMap<std::__1::set<Standalone<StringRef>, std::__1::less<Standalone<StringRef> >, std::__1::allocator<Standalone<StringRef> > >, int, ConstantMetric<int> >*, KeyRangeMap<ServerCacheInfo, int, ConstantMetric<int> >*, std::__1::map<Standalone<StringRef>, applyMutationsData, std::__1::less<Standalone<StringRef> >, std::__1::allocator<std::__1::pair<Standalone<StringRef> const, applyMutationsData> > >*, RequestStream<CommitTransactionRequest>, Database, NotifiedVersion*, std::__1::map<UID, Reference<StorageInfo>, std::__1::less<UID>, std::__1::allocator<std::__1::pair<UID const, Reference<StorageInfo> > > >*, std::__1::map<Tag, long, std::__1::less<Tag>, std::__1::allocator<std::__1::pair<Tag const, long> > >*, bool))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::value, TraceEvent&>::type TraceEvent::detail<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<UID>::value, TraceEvent&>::type TraceEvent::detail<UID>(char const*, UID const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<Standalone<StringRef> >::value, TraceEvent&>::type TraceEvent::detail<Standalone<StringRef> >(char const*, Standalone<StringRef> const&))

ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by vector:1010 (/usr/local/include/c++/v1/vector:1010)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(LogPushData::addTag(Tag))

ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by __split_buffer:0 (/usr/local/include/c++/v1/__split_buffer:0)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(void LogPushData::addTypedMessage<MutationRef>(MutationRef const&, bool))

ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by __split_buffer:0 (/usr/local/include/c++/v1/__split_buffer:0)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(void LogPushData::addTypedMessage<LogProtocolMessage>(LogProtocolMessage const&, bool))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<StringRef>::value, TraceEvent&>::type TraceEvent::detail<StringRef>(char const*, StringRef const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<bool>::value, TraceEvent&>::type TraceEvent::detail<bool>(char const*, bool const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<int>::value, TraceEvent&>::type TraceEvent::detail<int>(char const*, int const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<long>::value, TraceEvent&>::type TraceEvent::detail<long>(char const*, long const&))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<char [36]>::value, TraceEvent&>::type TraceEvent::detail<char [36]>(char const*, char const (&) [36]))

ld.lld: error: undefined symbol: std::__1::__basic_string_common<true>::__throw_length_error() const
>>> referenced by string:1784 (/usr/local/include/c++/v1/string:1784)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(std::__1::enable_if<Traceable<char [5]>::value, TraceEvent&>::type TraceEvent::detail<char [5]>(char const*, char const (&) [5]))

ld.lld: error: undefined symbol: std::__1::__vector_base_common<true>::__throw_length_error() const
>>> referenced by vector:976 (/usr/local/include/c++/v1/vector:976)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(ServerCacheInfo::ServerCacheInfo(ServerCacheInfo const&))

ld.lld: error: undefined symbol: std::__1::error_category::~error_category()
>>> referenced by error_code.hpp:228 (/opt/boost_1_67_0/boost/system/error_code.hpp:228)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(boost::system::error_category::std_category::~std_category())

ld.lld: error: undefined symbol: std::__1::generic_category()
>>> referenced by error_code.hpp:702 (/opt/boost_1_67_0/boost/system/error_code.hpp:702)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(boost::system::error_category::std_category::equivalent(int, std::__1::error_condition const&) const)

ld.lld: error: undefined symbol: typeinfo for std::__1::error_category
>>> referenced by error_code.hpp:711 (/opt/boost_1_67_0/boost/system/error_code.hpp:711)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(boost::system::error_category::std_category::equivalent(int, std::__1::error_condition const&) const)

ld.lld: error: undefined symbol: std::__1::generic_category()
>>> referenced by error_code.hpp:732 (/opt/boost_1_67_0/boost/system/error_code.hpp:732)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(boost::system::error_category::std_category::equivalent(std::__1::error_code const&, int) const)

ld.lld: error: undefined symbol: typeinfo for std::__1::error_category
>>> referenced by error_code.hpp:741 (/opt/boost_1_67_0/boost/system/error_code.hpp:741)
>>>               .objs/fdbserver/ApplyMetadataMutation.cpp.o:(boost::system::error_category::std_category::equivalent(std::__1::error_code const&, int) const)

ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [bin/fdbserver] Error 1

```

Reply:
My comment is that there are 4 configurations, and this only allows 2 of them to exist.  In particular, clang+libstdc++ would be a very useful configuration for us as it would let us actually use clang and not produce potentially broken binaries, and it'd be... really nice to (a) let that exist and (b) look into why it doesn't work a bit deeper.

Reply:
@jzhou77 According to google search results, that issue goes away if you link using $CXX instead of $CC in build/link-wrapper.sh

Reply:
I tried with $CXX, but still the same error. My invoke command is:

```
/usr/local/bin/clang++.b0bd9fc1 -v -fuse-ld=lld -Wl,--disable-new-dtags .objs/fdbserver/ApplyMetadataMutation.cpp.o .objs/fdbserver/Knobs.cpp.o .objs/fdbserver/LatencyBandConfig.cpp.o .objs/fdbserver/SkipList.cpp.o .objs/fdbserver/VFSAsync.cpp.o .objs/fdbserver/sqlite/sqlite3.amalgamation.c.o .objs/fdbserver/workloads/AsyncFile.cpp.o .objs/fdbserver/workloads/Fuzz.cpp.o .objs/fdbserver/workloads/MemoryKeyValueStore.cpp.o .objs/fdbserver/ClusterController.actor.g.cpp.o .objs/fdbserver/CoordinatedState.actor.g.cpp.o .objs/fdbserver/Coordination.actor.g.cpp.o .objs/fdbserver/CoroFlow.actor.g.cpp.o .objs/fdbserver/DataDistribution.actor.g.cpp.o .objs/fdbserver/DataDistributionQueue.actor.g.cpp.o .objs/fdbserver/DataDistributionTracker.actor.g.cpp.o .objs/fdbserver/DiskQueue.actor.g.cpp.o .objs/fdbserver/FDBExecHelper.actor.g.cpp.o .objs/fdbserver/IndirectShadowPager.actor.g.cpp.o .objs/fdbserver/KeyValueStoreCompressTestData.actor.g.cpp.o .objs/fdbserver/KeyValueStoreMemory.actor.g.cpp.o .objs/fdbserver/KeyValueStoreSQLite.actor.g.cpp.o .objs/fdbserver/LeaderElection.actor.g.cpp.o .objs/fdbserver/LogRouter.actor.g.cpp.o .objs/fdbserver/LogSystemDiskQueueAdapter.actor.g.cpp.o .objs/fdbserver/LogSystemPeekCursor.actor.g.cpp.o .objs/fdbserver/MasterProxyServer.actor.g.cpp.o .objs/fdbserver/MemoryPager.actor.g.cpp.o .objs/fdbserver/MoveKeys.actor.g.cpp.o .objs/fdbserver/OldTLogServer_4_6.actor.g.cpp.o .objs/fdbserver/OldTLogServer_6_0.actor.g.cpp.o .objs/fdbserver/QuietDatabase.actor.g.cpp.o .objs/fdbserver/Ratekeeper.actor.g.cpp.o .objs/fdbserver/Resolver.actor.g.cpp.o .objs/fdbserver/Restore.actor.g.cpp.o .objs/fdbserver/SimulatedCluster.actor.g.cpp.o .objs/fdbserver/Status.actor.g.cpp.o .objs/fdbserver/TLogServer.actor.g.cpp.o .objs/fdbserver/TagPartitionedLogSystem.actor.g.cpp.o .objs/fdbserver/VersionedBTree.actor.g.cpp.o .objs/fdbserver/WaitFailure.actor.g.cpp.o .objs/fdbserver/fdbserver.actor.g.cpp.o .objs/fdbserver/masterserver.actor.g.cpp.o .objs/fdbserver/networktest.actor.g.cpp.o .objs/fdbserver/pubsub.actor.g.cpp.o .objs/fdbserver/storageserver.actor.g.cpp.o .objs/fdbserver/tester.actor.g.cpp.o .objs/fdbserver/worker.actor.g.cpp.o .objs/fdbserver/workloads/ApiCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/ApiWorkload.actor.g.cpp.o .objs/fdbserver/workloads/AsyncFileCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/AsyncFileRead.actor.g.cpp.o .objs/fdbserver/workloads/AsyncFileWrite.actor.g.cpp.o .objs/fdbserver/workloads/AtomicOps.actor.g.cpp.o .objs/fdbserver/workloads/AtomicOpsApiCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/AtomicRestore.actor.g.cpp.o .objs/fdbserver/workloads/AtomicSwitchover.actor.g.cpp.o .objs/fdbserver/workloads/BackgroundSelectors.actor.g.cpp.o .objs/fdbserver/workloads/BackupCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/BackupToDBAbort.actor.g.cpp.o .objs/fdbserver/workloads/BackupToDBCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/BackupToDBUpgrade.actor.g.cpp.o .objs/fdbserver/workloads/BulkLoad.actor.g.cpp.o .objs/fdbserver/workloads/ChangeConfig.actor.g.cpp.o .objs/fdbserver/workloads/ClientTransactionProfileCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/CommitBugCheck.actor.g.cpp.o .objs/fdbserver/workloads/ConfigureDatabase.actor.g.cpp.o .objs/fdbserver/workloads/ConflictRange.actor.g.cpp.o .objs/fdbserver/workloads/ConsistencyCheck.actor.g.cpp.o .objs/fdbserver/workloads/CpuProfiler.actor.g.cpp.o .objs/fdbserver/workloads/Cycle.actor.g.cpp.o .objs/fdbserver/workloads/DDBalance.actor.g.cpp.o .objs/fdbserver/workloads/DDMetrics.actor.g.cpp.o .objs/fdbserver/workloads/DDMetricsExclude.actor.g.cpp.o .objs/fdbserver/workloads/DifferentClustersSameRV.actor.g.cpp.o .objs/fdbserver/workloads/DiskDurability.actor.g.cpp.o .objs/fdbserver/workloads/DiskDurabilityTest.actor.g.cpp.o .objs/fdbserver/workloads/DummyWorkload.actor.g.cpp.o .objs/fdbserver/workloads/ExternalWorkload.actor.g.cpp.o .objs/fdbserver/workloads/FastTriggeredWatches.actor.g.cpp.o .objs/fdbserver/workloads/FileSystem.actor.g.cpp.o .objs/fdbserver/workloads/FuzzApiCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/Increment.actor.g.cpp.o .objs/fdbserver/workloads/IndexScan.actor.g.cpp.o .objs/fdbserver/workloads/Inventory.actor.g.cpp.o .objs/fdbserver/workloads/KVStoreTest.actor.g.cpp.o .objs/fdbserver/workloads/KillRegion.actor.g.cpp.o .objs/fdbserver/workloads/LocalRatekeeper.actor.g.cpp.o .objs/fdbserver/workloads/LockDatabase.actor.g.cpp.o .objs/fdbserver/workloads/LogMetrics.actor.g.cpp.o .objs/fdbserver/workloads/LowLatency.actor.g.cpp.o .objs/fdbserver/workloads/MachineAttrition.actor.g.cpp.o .objs/fdbserver/workloads/Mako.actor.g.cpp.o .objs/fdbserver/workloads/MemoryLifetime.actor.g.cpp.o .objs/fdbserver/workloads/MetricLogging.actor.g.cpp.o .objs/fdbserver/workloads/Performance.actor.g.cpp.o .objs/fdbserver/workloads/Ping.actor.g.cpp.o .objs/fdbserver/workloads/PubSubMultiples.actor.g.cpp.o .objs/fdbserver/workloads/QueuePush.actor.g.cpp.o .objs/fdbserver/workloads/RYWDisable.actor.g.cpp.o .objs/fdbserver/workloads/RYWPerformance.actor.g.cpp.o .objs/fdbserver/workloads/RandomClogging.actor.g.cpp.o .objs/fdbserver/workloads/RandomMoveKeys.actor.g.cpp.o .objs/fdbserver/workloads/RandomSelector.actor.g.cpp.o .objs/fdbserver/workloads/ReadWrite.actor.g.cpp.o .objs/fdbserver/workloads/RemoveServersSafely.actor.g.cpp.o .objs/fdbserver/workloads/Rollback.actor.g.cpp.o .objs/fdbserver/workloads/RyowCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/SaveAndKill.actor.g.cpp.o .objs/fdbserver/workloads/SelectorCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/Serializability.actor.g.cpp.o .objs/fdbserver/workloads/Sideband.actor.g.cpp.o .objs/fdbserver/workloads/SlowTaskWorkload.actor.g.cpp.o .objs/fdbserver/workloads/SnapTest.actor.g.cpp.o .objs/fdbserver/workloads/StatusWorkload.actor.g.cpp.o .objs/fdbserver/workloads/Storefront.actor.g.cpp.o .objs/fdbserver/workloads/StreamingRead.actor.g.cpp.o .objs/fdbserver/workloads/SuspendProcesses.actor.g.cpp.o .objs/fdbserver/workloads/TargetedKill.actor.g.cpp.o .objs/fdbserver/workloads/TaskBucketCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/ThreadSafety.actor.g.cpp.o .objs/fdbserver/workloads/Throttling.actor.g.cpp.o .objs/fdbserver/workloads/Throughput.actor.g.cpp.o .objs/fdbserver/workloads/TimeKeeperCorrectness.actor.g.cpp.o .objs/fdbserver/workloads/TriggerRecovery.actor.g.cpp.o .objs/fdbserver/workloads/UnitPerf.actor.g.cpp.o .objs/fdbserver/workloads/UnitTests.actor.g.cpp.o .objs/fdbserver/workloads/Unreadable.actor.g.cpp.o .objs/fdbserver/workloads/VersionStamp.actor.g.cpp.o .objs/fdbserver/workloads/WatchAndWait.actor.g.cpp.o .objs/fdbserver/workloads/Watches.actor.g.cpp.o .objs/fdbserver/workloads/WorkerErrors.actor.g.cpp.o .objs/fdbserver/workloads/WriteBandwidth.actor.g.cpp.o .objs/fdbserver/workloads/WriteDuringRead.actor.g.cpp.o lib/libfdbclient.a lib/libfdbrpc.a lib/libflow.a lib/libFDBLibTLS.a -Bstatic /usr/local/lib/libtls.a /usr/local/lib/libssl.a /usr/local/lib/libcrypto.a -Bdynamic -ldl -lpthread -lrt -static-libgcc -o bin/fdbserver /usr/lib/gcc/x86_64-redhat-linux/4.4.7/libstdc++.a -lm
```